### PR TITLE
pr - 005 - Even more error propagation

### PR DESCRIPTION
**Pull request by jxsl13**
This is the breaking subset of changes... ~~will analyze after some sleep~~

Seemingly there are actually messages that are skipped which the server sends.
My error handling incorrectly introduced errors instead of the returned bool values.

In the specific edge case, where we want to skip processing instead of actually error due to some unknown message type, the previous implementation returned an error instead of skipping processing.

This implementation now correctly propagates errors from within callbacks as well as skips unknown message types.

### CODE REVIEW

**Comment by jxsl13** [comment]
there is still a small edge case which I need to take care of in internal error handling

File: teeworlds7/packet.go [hunk 1]
```diff
@@ -48,23 +54,25 @@ func (client *Client) processPacket(packet *protocol7.Packet) error {
 
 	if packet.Header.Flags.Control {
 		if len(packet.Messages) != 1 {
-			return fmt.Errorf("got control packet with %d messages.\n", len(packet.Messages))
+			return fmt.Errorf("got control packet with %d messages: expected only 1", len(packet.Messages))
```
Comment (ChillerDragon) [inline teeworlds7/packet.go:57]: One or zero. There can be empty control messages if a resend is requested.

File: examples/client_simple/main.go [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+func main() {
+	ctx, cancel := signal.NotifyContext(context.Background(), os.Interrupt, syscall.SIGTERM)
+	defer cancel()
+
+	client := teeworlds7.NewClient()
+	client.Name = "nameless tee"
+
+	client.OnPacket(func(packet *protocol7.Packet) bool {
+		fmt.Printf("received packet w/ %d messages\n", len(packet.Messages))
```
Comment (ChillerDragon) [inline examples/client_simple/main.go:23]: This seems a bit verbose for a "client_simple" given that there is already a "client_verbose"

Also I personally am not a big fan of "w/" ._.

File: teeworlds7/send_message_hooks.go [hunk 3]
```diff
@@ -19,306 +19,307 @@ func userSendMsgCallback[T any](userCallbacks []func(T) bool, msg T) bool {
...
-				continue
+			if !callback(&msg) {
+				continue outer
 			}
 		}
 
 		switch castMsg := msg.(type) {
 		case *messages7.CtrlKeepAlive:
-			if userSendMsgCallback(client.Callbacks.CtrlKeepAliveOut, castMsg) == false {
+			if !userSendMsgCallback(client.Callbacks.CtrlKeepAliveOut, castMsg) {
```
Comment (ChillerDragon) [inline teeworlds7/send_message_hooks.go:32]: Stuff like that should be on its own pr. I typed out all the ``== false`` because I thought its more the go vibe. But I am totally fine with the ``!`` I use that everywhere else.

You can leave it in for now. I just wanted to say that a huge opinionated wall of style diff does not belong into a error propagation pr. Makes it harder to scroll through the diff and discuss what is relevant.

Comment (jxsl13) [reply teeworlds7/send_message_hooks.go:32]: ` == true/false` is not Go style.
We got a linter for stuff like this, which is why I'm conforming to "staticcheck" warnings.
You can easily install it if you got the vscode Go extension
Bildschirmfoto 2024-09-12 um 18 10 07
Bildschirmfoto 2024-09-12 um 18 12 16

File: examples/client_simple/main.go (hunk 2 above)
Comment (jxsl13) [reply examples/client_simple/main.go:23]: resolved

File: teeworlds7/send_message_hooks.go (hunk 3 above)
Comment (jxsl13) [reply teeworlds7/send_message_hooks.go:32]: There is no style discussion in Go, as most of the expected styling is documented on the Go website. These discussions are a waste of time.

File: teeworlds7/packet.go (hunk 1 above)
Comment (jxsl13) [reply teeworlds7/packet.go:57]: does this mean that the logic was previously incorrect? because the check  above did explicitly check whether it's exactly one element.

Comment (jxsl13) [reply teeworlds7/packet.go:57]: What's the best way to fix this :?

Comment (ChillerDragon) [reply teeworlds7/packet.go:57]: The best way to fix this is to implement resends and correctly read the flags even for empty packets.
But that is out of scope for this pr.

Until then let's just not touch the error message.

File: examples/client_simple/main.go [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+func main() {
+	ctx, cancel := signal.NotifyContext(context.Background(), os.Interrupt, syscall.SIGTERM)
+	defer cancel()
+
+	client := teeworlds7.NewClient()
+	client.Name = "nameless tee"
+
+	client.OnPacket(func(packet *protocol7.Packet) bool {
+		fmt.Printf("received packet with %d message(s)\n", len(packet.Messages))
```
Comment (ChillerDragon) [inline examples/client_simple/main.go:23]: This is still printing every message in the "client_simple" example. Seems quite redundant to the "client_verbose" maybe you could name it differently to describe what this example is about. Or merge them together. Or make it not spam print.

File: teeworlds7/send_message_hooks.go (hunk 3 above)
Comment (ChillerDragon) [reply teeworlds7/send_message_hooks.go:32]: Ah nice to hear that a linter agrees.

File: examples/client_simple/main.go (hunk 2 above)
Comment (jxsl13) [reply examples/client_simple/main.go]: reverted message, removed newline & punctuation.

File: teeworlds7/packet.go (hunk 1 above)
Comment (jxsl13) [reply teeworlds7/packet.go:57]: reverted message, removed newline & punctuation.

File: examples/client_simple/main.go (hunk 4 above)
Comment (jxsl13) [reply examples/client_simple/main.go]: changed code a little bit to actually fit some specific use case and renamed the example application to "client_echo".

**Comment by ChillerDragon** [comment]
Thanks
